cloudant-sync-datastore-core/src/main/java/com/cloudant/sync/internal/replication/ReplicationListener.java

```diff
@@ -0,0 +1,75 @@
+package com.cloudant.sync.internal.replication;
+
+import com.cloudant.sync.event.Subscribe;
+import com.cloudant.sync.event.notifications.ReplicationCompleted;
+import com.cloudant.sync.event.notifications.ReplicationErrored;
+import com.cloudant.sync.replication.ReplicationPolicyManager;
+import com.cloudant.sync.replication.Replicator;
+
+import java.util.HashSet;
+import java.util.Set;
+
+/**
+ * This class is not intended as API, it is public for EventBus access only.
+ * API consumers should not call these methods. See #setReplicationsCompletedListener for more
+ * information on replication lifecycle events.
+ *
+ * @api_private
+ */
+public class ReplicationListener {
+
+    private final Set<Replicator> replicatorsInProgress = new HashSet<Replicator>();
+    private ReplicationPolicyManager.ReplicationsCompletedListener mReplicationsCompletedListener
+            = null;
+
+    public void setReplicationsCompletedListener(ReplicationPolicyManager
+                                                         .ReplicationsCompletedListener
+                                                         mReplicationsCompletedListener) {
+        this.mReplicationsCompletedListener = mReplicationsCompletedListener;
+    }
+
+    public void add(Replicator replicator) {
+        synchronized (replicatorsInProgress) {
+            replicatorsInProgress.add(replicator);
+        }
+    }
+
+    public void remove(Replicator replicator) {
+        synchronized (replicatorsInProgress) {
+            if (replicatorsInProgress.remove(replicator)) {
+                replicator.getEventBus().unregister(this);
```

Comment:
As discussed there might be some subtle issues around eventBus registration and de-registration